node-wasm/src/ports.rs

```diff
@@ -0,0 +1,286 @@
+use js_sys::{Array, Function, Reflect};
+use serde::Serialize;
+use serde_wasm_bindgen::{from_value, to_value, Serializer};
+use tokio::select;
+use tokio::sync::{mpsc, Mutex};
+use tracing::{error, info, trace};
+use wasm_bindgen::closure::Closure;
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::{JsCast, JsValue};
+use web_sys::{MessageEvent, MessagePort};
+
+use crate::commands::{NodeCommand, WorkerResponse};
+use crate::error::{Context, Error, Result};
+use crate::utils::MessageEventExt;
+
+// Instead of supporting communication with just `MessagePort`, allow using any object which
+// provides compatible interface, eg. `Worker`
+#[wasm_bindgen]
+extern "C" {
+    pub type MessagePortLike;
+
+    #[wasm_bindgen(catch, method, structural, js_name = postMessage)]
+    pub fn post_message(this: &MessagePortLike, message: &JsValue) -> Result<(), JsValue>;
+
+    #[wasm_bindgen(catch, method, structural, js_name = postMessage)]
+    pub fn post_message_with_transferable(
+        this: &MessagePortLike,
+        message: &JsValue,
+        transferable: &JsValue,
+    ) -> Result<(), JsValue>;
+}
+
+impl From<MessagePort> for MessagePortLike {
+    fn from(port: MessagePort) -> Self {
+        JsValue::from(port).into()
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct ClientId(usize);
+
+struct ClientConnection {
+    port: MessagePortLike,
+    _onmessage: Closure<dyn Fn(MessageEvent)>,
+}
+
+impl ClientConnection {
+    fn new(
+        id: ClientId,
+        object: JsValue,
+        forward_messages_to: mpsc::UnboundedSender<(ClientId, NodeCommand)>,
+        forward_connects_to: mpsc::UnboundedSender<JsValue>,
```

Comment:
Since both sender are constructed from `WorkerServer` and their receivers are handled in the same function. Isn't it better to have a single enum for them?

Reply:
Good idea, switched to single channel :+1: